### PR TITLE
tests/candev: Prevent user from attempting to use native can interface on non-native board

### PR DESCRIPTION
**Pull request by wosym**
### Contribution description

When a user tries to use tests/candev on a non-native board without specifying a can_driver (or making a typo while specifying one, like I did yesterday), the test app will default to the native Linux driver. Of course, this is fine when running the test-app on native, but using Linux drivers on (for example) stm32... that of course will not work. 
In the current form, the test app will compile without any warnings and also start without any issues. However, it will immediately go into a kernel-panic when the driver is loaded. 

If you're unfamiliar with the expected behavior of this test app this can be quite confusing, so I propose to prevent the user from making this mistake. 

### Testing procedure

* on master: 
`make all flash term BOARD=nucleo-f207zg`
--> will compile without any warnings, but will throw the following output immediately after resetting the board:
```
2020-08-22 12:13:26,564 # 0x8000eaf
2020-08-22 12:13:26,565 # *** RIOT kernel panic:
2020-08-22 12:13:26,565 # FAILED ASSERTION.
2020-08-22 12:13:26,565 # 
2020-08-22 12:13:26,566 # *** halted.
```

* with this patch:
`make all flash term BOARD=nucleo-f207zg`
compiling will fail and tell you immediately what's wrong:
```
Makefile:13: *** native can driver can only be used on native board!.  Stop.
```

### CODE REVIEW

File: tests/candev/Makefile
Comment (leandrolanzieri) [inline tests/candev/Makefile:18]:
```suggestion
ifeq ($(CAN_DRIVER), native)
  ifneq ($(BOARD), native)
    $(error native can driver can only be used on native board!)
  endif
endif

```

File: tests/candev/Makefile
Comment (leandrolanzieri) [inline tests/candev/Makefile:21]: Question: the check bellow is supposed to match the default `native` value of `CAN_DRIVER`?

**Review by leandrolanzieri — COMMENTED** [review]
Makes sense, just a small style comment.

File: tests/candev/Makefile
Comment (wosym) [reply tests/candev/Makefile:21]: I was confused about what you were asking here at first, but then it dawned on me: we set the default value to `native`, but check for a set define called `CAN_NATIVE`. Good catch! 

I suggest changing the default value though, so it follows the convention.

File: tests/candev/Makefile
Comment (leandrolanzieri) [inline tests/candev/Makefile:11]: Now that this is in place, the CI should only attempt to build for `native` right? Otherwise it would fail.

Comment (leandrolanzieri) [reply tests/candev/Makefile:11]: Wouldn't it be possible to set the driver according to the used board and use the test for all of them?

Comment (wosym) [reply tests/candev/Makefile:11]: Mhmmm, yes. Because the Makefile will throw that error, the CI will fail indeed. Unless... the CI selects a can_driver when building for non-native. 
Could you guide me how I best solve this?

Comment (wosym) [reply tests/candev/Makefile:11]: Setting the driver according to the used board is not possible because because there is no correlation between board and driver. You can use any driver you want (as long as it's implemented in RIOT already) with any board you want.

Comment (leandrolanzieri) [reply tests/candev/Makefile:11]: Taking a second look at the test, what is this actually testing when not running on native? For any board that is not native it will assert in the main function. Is it worth compiling this application for all boards in the CI? If not, the following can be added to the Makefile:

```Makefile
BOARD_WHITELIST := native
```

**Comment by leandrolanzieri** [comment]
Another question: why is this CFLAG being injected in the Makefile?
`CFLAGS += -DCAN_DRIVER_$(CAN_DRIVER)`

I can't find any code using it.

File: tests/candev/Makefile
Comment (wosym) [reply tests/candev/Makefile:11]: > Is it worth compiling this application for all boards in the CI? If

To be honest I don't think it is useful, no. I'll add that line to the Makefile. It will not have any influence on making locally, right?

**Comment by wosym** [comment]
> Another question: why is this CFLAG being injected in the Makefile?
> `CFLAGS += -DCAN_DRIVER_$(CAN_DRIVER)`
> 
> I can't find any code using it.

Aha, it looks weird in current form indeed, because for native we don't check for `CAN_DRIVER_native` but for `BOARD_NATIVE`.
It makes it look kinda useless, but it should be used when initializing other can drivers than native. See line 200 in main.c
For now, that spot is still empty, but in [this](https://github.com/RIOT-OS/RIOT/pull/13624) PR it will be filled in.

But I guess renaming it will clear up confusion in the meantime.

File: tests/candev/Makefile
Comment (leandrolanzieri) [reply tests/candev/Makefile:11]: >  It will not have any influence on making locally, right?

It would prevent local building as well, unless `CONTINUE_ON_EXPECTED_ERRORS=1` is set.

>(btw. I assume you meant adding it to Makefile.ci, not Makefile. Right?)

Actually I meant the `Makefile`, for some reason be place the list in that file also for other tests.

File: tests/candev/main.c
Comment (leandrolanzieri) [inline tests/candev/main.c:195]: I think It would be better to use the automatic generated macros by the build system for the modules. Instead of creating your own CFLAG I would check:
```suggestion
#if IS_USED(MODULE_CAN_LINUX)
```

Then it can be extended for the other drivers when introduced.

File: tests/candev/Makefile
Comment (wosym) [reply tests/candev/Makefile:11]: > It would prevent local building as well, unless CONTINUE_ON_EXPECTED_ERRORS=1 is set.

Mhmmm, but... then why would we want to include this? 

> Actually I meant the Makefile, for some reason be place the list in that file also for other tests.

Oh, Ok. I will fix it once the above has been clarified.

File: tests/candev/main.c
Comment (wosym) [reply tests/candev/main.c]: Not sure how it will work out in the future (I will reevaluate the change in the next PR. If it makes things a lot more complicated I might consider reverting the change).

But I applied your suggestion for now.

File: tests/candev/Makefile
Comment (leandrolanzieri) [reply tests/candev/Makefile:11]: > Mhmmm, but... then why would we want to include this?

So the CI does not attempt to build the test for boards other than native.

Comment (wosym) [reply tests/candev/Makefile:11]: But you said this would also prevent users from building for non-native boards locally on their machines?

Comment (leandrolanzieri) [reply tests/candev/Makefile:11]: > But you said this would also prevent users from building for non-native boards locally on their machines?

Yes, unless when compiling using `CONTINUE_ON_EXPECTED_ERRORS=1` (and of course overriding `CAN_DRIVER` so the check introduced here does not raise an error). What is the use case for running this test (as is) on a board that is not native?

Comment (wosym) [reply tests/candev/Makefile:11]: Are you suggesting that the user should always add `CONTINUE_ON_EXPECTED_ERRORS=1` himself? That doesn't feel like it's a good solution? I would expect it to work out of the box? 
User picks a board, picks a driver, and without any changes it works. 

> What is the use case for running this test (as is) on a board that is not native?

as is, not much. You will have to select a non-native driver. Either you implement the code for PERIPH_CAN (such as stm32 has), or you use the driver from the MCP2515 PR.

Comment (leandrolanzieri) [reply tests/candev/Makefile:11]: > User picks a board, picks a driver, and without any changes it works.

Yes, I would expect that as well, which is not possible with the current state of the test, as it needs native implementation to work, or code modification in main to initialize the chosen driver. The whitelist can be removed once this limitation is not there anymore.

I don't really see the goal of this CFLAG when it is actually not pulling any driver module in...I would expect (when more drivers are there) to set this Makefile variable and that to be added to `USEMODULE`, like so:
https://github.com/RIOT-OS/RIOT/blob/7eed33bf7870a5e535d60f217acd70ac9cc05c0f/tests/driver_sx127x/Makefile#L10-L13

Comment (wosym) [reply tests/candev/Makefile:11]: Ok, I think I get what you're saying now. I added `BOARD_WHITELIST := native` to the makefile now.

> I don't really see the goal of this CFLAG when it is actually not pulling any driver module in
I removed it a few patches ago.

**Review by leandrolanzieri — APPROVED** [review]
ACK.

**Comment by fjmolinas** [comment]
This error currently make `compile_and_test_for_board` exit because of a command error. Why the need for the make error? there is already a `BOARD_WHITELIST`, if someone tries to compile this they already get an error.

Why not add a `FEATURE` for this? Then `can` could do `FEATURES_REQUIRED_ANY += periph_can|can_linux`.

IMO we should only leave the whitelist in the test, the rest doesn't make sense.

**Comment by leandrolanzieri** [comment]
Yes, you are right, in the current state of the test we can drop `CAN_DEVICE` altogether and just keep the whitelist. It does not change the included modules and is not checked by the code. The test works only for native currently anyways.

**Comment by wosym** [comment]
@fjmolinas and @leandrolanzieri : What changes do you propose precisely? Creating a separate PR for them probably is not worth it, but I can incorporate them in  #13624 if neccesary.  

The CAN_DRIVER define is quite important though, because it will determine what can-driver modules are included and how the initialization process will work.

**Comment by fjmolinas** [comment]
> @fjmolinas and @leandrolanzieri : What changes do you propose precisely? Creating a separate PR for them probably is not worth it, but I can incorporate them in #13624 if neccesary.

I would like to remove the compile error. Currently this is not how we handle tests with insufficient `FEATURES`. With the `BOARD_WHITELIST` this is already addressed (I removed the make error)

```
make: Entering directory '/home/francisco/workspace/RIOT/tests/candev'
The selected BOARD=nucleo-l152re is not whitelisted: native
/home/francisco/workspace/RIOT/tests/candev/../../Makefile.include:908: *** You can let the build continue on expected errors by setting CONTINUE_ON_EXPECTED_ERRORS=1 to the command line.  Stop.
make: Leaving directory '/home/francisco/workspace/RIOT/tests/candev'
```

The rest of the logic at the time of this PR was not needed. With the error `info-board-supported` does not work for the test:

```
make: Entering directory '/home/francisco/workspace/RIOT/tests/candev'
Makefile:14: *** native can driver can only be used on native board!.  Stop.
make: Leaving directory '/home/francisco/workspace/RIOT/tests/candev'
```

For me the `$(error native can driver can only be used on native board!)` needs to be removed ASAP because of this.

> The CAN_DRIVER define is quite important though, because it will determine what can-driver modules are included and how the initialization process will work.

How exactly? When I grep for `CAN_DRIVER` I only get

```
→ git grep CAN_NATIVE
tests/candev/Makefile:CAN_DRIVER ?= CAN_NATIVE
tests/candev/Makefile:ifeq ($(CAN_DRIVER), CAN_NATIVE)
tests/candev/Makefile:else ifeq ($(CAN_DRIVER), CAN_NATIVE)
```

If #13624 needs these then it should have been introduced there and not here.

**Comment by wosym** [comment]
> I would like to remove the compile error. Currently this is not how we handle tests with insufficient `FEATURES`. With the `BOARD_WHITELIST` this is already addressed (I removed the make error)
> ```
> make: Entering directory '/home/francisco/workspace/RIOT/tests/candev'
> The selected BOARD=nucleo-l152re is not whitelisted: native
> /home/francisco/workspace/RIOT/tests/candev/../../Makefile.include:908: *** You can let the build continue on expected errors by setting CONTINUE_ON_EXPECTED_ERRORS=1 to the command line.  Stop.
> make: Leaving directory '/home/francisco/workspace/RIOT/tests/candev'
> ```

So... the BOARD_WHITELIST prevents the app from being used on anything but native. Ok. But with the future in mind: this will have to be removed again, because it is meant to be used on non-native. What method do you propose to prevent the user from using native can drivers on an mcu?


> 
> The rest of the logic at the time of this PR was not needed. With the error `info-board-supported` does not work for the test:
> 
> ```
> make: Entering directory '/home/francisco/workspace/RIOT/tests/candev'
> Makefile:14: *** native can driver can only be used on native board!.  Stop.
> make: Leaving directory '/home/francisco/workspace/RIOT/tests/candev'
> ```
> 
> For me the `$(error native can driver can only be used on native board!)` needs to be removed ASAP because of this.

Ok, I don't mind removing it. Since you say "ASAP", does that mean you want me to remove it in a separate PR? Or wait for #13624 to be merged? 

> > The CAN_DRIVER define is quite important though, because it will determine what can-driver modules are included and how the initialization process will work.
> 
> How exactly? When I grep for `CAN_DRIVER` I only get
> 
> ```
> → git grep CAN_NATIVE
> tests/candev/Makefile:CAN_DRIVER ?= CAN_NATIVE
> tests/candev/Makefile:ifeq ($(CAN_DRIVER), CAN_NATIVE)
> tests/candev/Makefile:else ifeq ($(CAN_DRIVER), CAN_NATIVE)
> ```
> If #13624 needs these then it should have been introduced there and not here.

It is important to the test. not outside of it. It determines what modules have to be loaded and what init-functions have to be called. 
Yes, you could argue that this code should not have been introduced here, but only in a subsequent PR, but the idea was to provide a boilerplate for the user. He could use whatever driver he wanted, not just native or the MCP2515 driver. For example, there already is an stm32 can-driver in RIOT. If the user wanted, he could use that driver with this test. 
However, it was out of scope for that PR, since I did not have the necessary hardware to add the functionality myself. But even if I had, it still wasn't in the scope of that PR, because we deliberately split up the PR in two parts: 
1) Add native version of test app (with boilerplate to add other drivers than native)
2) Add MCP2515 driver and add it as an option to the candev testapp. 

A third step could be:
3) Add stm32 peripheral driver.

**Comment by fjmolinas** [comment]
@wosym BTW sorry for some of the terms I used: "doesn't make sense." was not very freindly.

> It is important to the test. not outside of it. It determines what modules have to be loaded and what init-functions have to be called.
> Yes, you could argue that this code should not have been introduced here, but only in a subsequent PR, but the idea was to provide a boilerplate for the user. He could use whatever driver he wanted, not just native or the MCP2515 driver. For example, there already is an stm32 can-driver in RIOT. If the user wanted, he could use that driver with this test.
> However, it was out of scope for that PR, since I did not have the necessary hardware to add the functionality myself. But even if I had, it still wasn't in the scope of that PR, because we deliberately split up the PR in two parts:
> 
>     Add native version of test app (with boilerplate to add other drivers than native)
>     Add MCP2515 driver and add it as an option to the candev testapp.
> 
> A third step could be:
> 3) Add stm32 peripheral driver.

Ok, I'm not aware of the whole PR process.

> Ok, I don't mind removing it. Since you say "ASAP", does that mean you want me to remove it in a separate PR? Or wait for #13624 to be merged?

I don't know is that PR soo to be merged? If it is then lets do it there! If not I think I might to a PR to remove the error (or turn it into a Warning), and to fix `compile_and_test_for_board` so it doesn't fail because of this. It basically depends on the timeline for that PR.

> So... the BOARD_WHITELIST prevents the app from being used on anything but native. Ok. But with the future in mind: this will have to be removed again, because it is meant to be used on non-native. What method do you propose to prevent the user from using native can drivers on an mcu?

I'll need to take a look at #13624 and `can` I haven't used it and I'm not aware of its state much. But is the test supposed to eventually work with any `can` "provider"?

**Comment by wosym** [comment]
> @wosym BTW sorry for some of the terms I used: "doesn't make sense." was not very freindly.
> 
No need to appologize! It takes a lot more for me to be offended ;) (To be honest I hadn't even paid attention to it...)


> 
> I don't know is that PR soo to be merged? If it is then lets do it there! If not I think I might to a PR to remove the error (or turn it into a Warning), and to fix `compile_and_test_for_board` so it doesn't fail because of this. It basically depends on the timeline for that PR.

I want to say "Yes", but I've thought that before... It has been postponed several times since I opened it in March (also due to my own delays: I've switched companies and have less time to work on it nowadays), so chances are that it'll still take a while for it to be merged. 
I can make a quick PR for it if you want. Shouldn't be too much work. Your call :) 


> 
> I'll need to take a look at #13624 and `can` I haven't used it and I'm not aware of its state much. But is the test supposed to eventually work with any `can` "provider"?

* Native only works on native. 
* MCP2515 is an external can driver controller over SPI. So unless there are other constraints (such as memory), you should be able to use this driver with any other board than native. Similar devices might exist. Same rule applies to all of them: should work with any board (besides native) as long as it has the require interface (spi, i2c, ...).
* Peripheral drivers (such as the stm32 can peripheral) will only work with the board (or cpu) it is associated with. So for example, the stm32 peripheral driver works with a nucleo-f207zg board.

**Comment by fjmolinas** [comment]
> MCP2515 is an external can driver controller over SPI. So unless there are other constraints (such as memory), you should be able to use this driver with any other board than native. Similar devices might exist. Same rule applies to all of them: should work with any board (besides native) as long as it has the require interface (spi, i2c, ...).

Ah damn, I've been wanting driver provided `FEATURES` for a while, maybe we could add that? But that would take a bit of time. Ok the variable makes sense in #13624. Could you add a `makefile.dep` in the test application where you do:

```
ifneq $(native, $(BOARD))
  USEMODULE += can_linux
else
  ifneq $(, $(filter mcp2515 $(USEMODULE))
    FEATURES_REQUIRED += periph_can
  endif
endif
```

Then if the user wants to explicitly use the driver the you add `USEMODULE=mcp2515` instead of `CANDEV_DRIVER` variable. Does that work?

But for now since the boiler plat for `PERIPH_CAN` is not in place, I think only the `BOARD_WHITELIST` is the best.

**Comment by wosym** [comment]
> Ah damn, I've been wanting driver provided `FEATURES` for a while, maybe we could add that? But that would take a bit of time. 
We could, but at the moment I'm not entirely grasping what you're suggesting. Feel free to elaborate. If I think it's feasible for that driver (either in a separate PR, or in the existing one), I might try to implement it.

> Ok the variable makes sense in #13624. Could you add a `makefile.dep` in the test application where you do:
Why a Makefile.dep? I check the other tests. They're not used for this kind of stuff? Shouldn't it be in the regular Makefile?

> ```
> ifneq $(native, $(BOARD))
>   USEMODULE += can_linux
> else
>   ifneq $(, $(filter mcp2515 $(USEMODULE))
>     FEATURES_REQUIRED += periph_can
>   endif
> endif
> ```
> 
> Then if the user wants to explicitly use the driver the you add `USEMODULE=mcp2515` instead of `CANDEV_DRIVER` variable. Does that work?

I think it should work, yes. However, for native this conditional can be dropped because it is implicitly included when you use the `can` module on native. That leaves us with only an else, that covers a not-implemented case.
Which... makes the whole conditional a bit silly... :)  (Even for a boilerplate)


> But for now since the boiler plat for `PERIPH_CAN` is not in place, I think only the `BOARD_WHITELIST` is the best.

The boilerplate will look quite different I think with the proposed changes above. I think it's best to make it a separate PR after all, to avoid confusion (including confusing myself...).

**Comment by fjmolinas** [comment]
@wosym I think we need to think about what the logic between the `can` modules is. So far I have the impression that the module `can` is basically the api definition and then `periph_can`, `can_linux`, and the new `mcp2515` implement that api. Is that correct?

If it is then we could think about having a `can` feature so a `FEATURES_PROVIDED += candev`. Then if the main `Makefile.dep`:

```
ifneq $(, $(filter can $(USEMODULE))
    FEATURES_REQUIRED +=candev|periph_can
endif
```

We then can add a `Makefile.features` for drivers and there in the `mcp2515` you would have a `Makefile.features` that would have:

```
FEATURES_PROVIDED += candev
```

Native should do the same:

```
ifeq ($(OS),Linux)
    FEATURES_PROVIDED += candev
endif

ifneq (,$(filter candev,$(FEATURES_USED)))
    USEMODULE += can_linux
endif
```

But this would mean a change in what we do currently since so far `MODULES` and or `DRIVERS` can't provide `FEATURES`, but this would be nice I think as long as there is some hardware behind the driver. But this would be a whole different PR and could not be liked by the community. And had already quite some discussion and initial proposals in #11676, #12952 and #12427. Also since we are migrating to Kconfig ATM, we need to keep that in mind as well. So this should be parallel work, not done inside #13624 (I'll make commetns in that PR for the rest).

> Why a Makefile.dep? I checked the other tests. They're not used for this kind of stuff? Shouldn't it be in the regular Makefile?

If its a `BOARD` dependency for an application/test then it should be in a `Makefile.board.dep` (theses do exist). Otherwise `info-boards-supported` yields different results depending on the intial value used for `BOARD` (there is some details on this in an old PR that I can't find..).

**Comment by fjmolinas** [comment]
> 

But this would mean a change in what we do currently since so far MODULES and or DRIVERS can't provide FEATURES, but this would be nice I think as long as there is some hardware behind the driver. But this would be a whole different PR and could not be liked by the community. And had already quite some discussion and initial proposals in #11676, #12952 and #12427. Also since we are migrating to Kconfig ATM, we need to keep that in mind as well. So this should be parallel work, not done inside #13624 (I'll make commetns in that PR for the rest).

The Kconfig part means that its probably not worth the effort to put more work in expanding the `FEATURES` system.

**Pull request merged**
